File: corelib/src/main/java/com/exmple/corelib/utils/CommonUtil.java
package com.exmple.corelib.utils;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Typeface;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.RequiresApi;
import android.support.v4.view.ViewCompat;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.style.StyleSpan;
import android.view.Display;
import android.view.KeyCharacterMap;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.GridView;
import android.widget.ListAdapter;
import android.widget.ListView;

import com.exmple.corelib.R;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by mou on 2016/7/6.
 */
public class CommonUtil {


    /**
     * 思路是先将statusbar设置成半透明，然后在contentView中添加一个和statusbar一样的高度的view作为statusbar
     * 的背景颜色
     *
     * @param activity
     */
    public static void customStatusbar(Activity activity) {
        Window window = activity.getWindow();
        ViewGroup mContentView = (ViewGroup) activity.findViewById(Window.ID_ANDROID_CONTENT);

        //First translucent status bar.
        window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        int statusBarHeight = getStatusBarHeight(activity);

        View mChildView = mContentView.getChildAt(0);
        if (mChildView != null) {
            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) mChildView.getLayoutParams();
            //如果已经为 ChildView 设置过了 marginTop, 再次调用时直接跳过
            if (lp != null && lp.topMargin < statusBarHeight && lp.height != statusBarHeight) {
                //不预留系统空间
                ViewCompat.setFitsSystemWindows(mChildView, false);
                lp.topMargin += statusBarHeight;
                mChildView.setLayoutParams(lp);
            }
        }
        // 已经添加过假背景view时，不用再次添加，只需要设置颜色即ok

        View statusBarView = mContentView.getChildAt(0);
        if (statusBarView != null && statusBarView.getLayoutParams() != null && statusBarView.getLayoutParams().height == statusBarHeight) {
            //避免重复调用时多次添加 View
            statusBarView.setBackgroundColor(activity.getResources().getColor(R.color.white));
//            statusBarView.setBackgroundResource(R.mipmap.ic_launcher);
            return;
        }
        statusBarView = new View(activity);
        ViewGroup.LayoutParams lp = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, statusBarHeight);
        statusBarView.setBackgroundColor(activity.getResources().getColor(R.color.white));
//        statusBarView.setBackgroundResource(R.mipmap.ic_launcher);
        //向 ContentView 中添加假 View
        mContentView.addView(statusBarView, 0, lp);
    }


    public static void fullScreen(Activity activity) {
        Window window = activity.getWindow();
        ViewGroup mContentView = (ViewGroup) activity.findViewById(Window.ID_ANDROID_CONTENT);

        //首先使 ChildView 不预留空间
        View mChildView = mContentView.getChildAt(0);
        if (mChildView != null) {
            ViewCompat.setFitsSystemWindows(mChildView, false);
        }

        int statusBarHeight = getStatusBarHeight(activity);
        //需要设置这个 flag 才能设置状态栏
        window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        //避免多次调用该方法时,多次移除了 View
        if (mChildView != null && mChildView.getLayoutParams() != null && mChildView.getLayoutParams().height == statusBarHeight) {
            //移除假的 View.
            mContentView.removeView(mChildView);
            mChildView = mContentView.getChildAt(0);
        }
        if (mChildView != null) {
            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) mChildView.getLayoutParams();
            //清除 ChildView 的 marginTop 属性
            if (lp != null && lp.topMargin >= statusBarHeight) {
//                lp.topMargin -= statusBarHeight;
                mChildView.setLayoutParams(lp);
            }
        }
    }

    public static int getStatusBarHeight(Activity activity) {
        int resId = activity.getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (resId > 0) {
            return activity.getResources().getDimensionPixelSize(resId);
        }
        return 0;
    }

    /**
     * 获取状态栏高度
     *
     * @param context
     * @return
     */
    public static int getNavigationBarHeight(Context context) {
        Resources resources = context.getResources();
        int resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android");
        return resources.getDimensionPixelSize(resourceId);
    }

    /**
     * 判断手机是否有导航栏
     *
     * @param context
     * @return
     */
    public static boolean hasNavigationBar(Context context) {
        //是否在屏幕外有菜单键
        boolean hasMenuKey = ViewConfiguration.get(context).hasPermanentMenuKey();
        //屏幕外是否有返回键
        boolean hasBackKey = KeyCharacterMap.deviceHasKey(KeyEvent.KEYCODE_BACK);
        return !hasBackKey && !hasMenuKey;

    }

    public static void startActivtiy(Context context, Class cls) {
        Intent intent = new Intent(context, cls);
        context.startActivity(intent);
    }

    public static void startActivtiy(Context context, Class cls, Bundle bundle) {
        if (bundle == null) {
            startActivtiy(context, cls);
            return;
        }
        Intent intent = new Intent(context, cls);
        intent.putExtras(bundle);
        context.startActivity(intent);
    }

    public static void startActivtiyForResult(Context context, Class cls, int requestCode) {
        Intent intent = new Intent(context, cls);
        if (context instanceof Activity) {
            ((Activity) context).startActivityForResult(intent, requestCode);
        } else {
            throw new IllegalThreadStateException();
        }
    }

    public static void startActivtiyForResult(Context context, Class cls, int requestCode, Bundle bundle) {
        if (bundle == null) {
            startActivtiyForResult(context, cls, requestCode);
            return;
        }
        Intent intent = new Intent(context, cls);
        intent.putExtras(bundle);
        if (context instanceof Activity) {
            ((Activity) context).startActivityForResult(intent, requestCode);
        } else {
            throw new IllegalThreadStateException();
        }
    }

    /**
     * 判断是否是电话号码
     *
     * @param phoneNum
     * @return
     */
    public static boolean isPhoneNum(String phoneNum) {
        if (TextUtils.isEmpty(phoneNum) || phoneNum.length() != 11) {
            return false;
        }
        return phoneNum.matches("[1][34578][0-9]{9}");
    }


    /**
     * Methods: setListViewHeight Description: 动态获取listview的高度
     *
     * @param listView void
     * @throws null
     */

    public static void setListViewHeight(ListView listView) {
        ListAdapter listAdapter = listView.getAdapter();
        if (listAdapter == null) {
            return;
        }
        int totalHeight = 0;
        for (int i = 0; i < listAdapter.getCount(); i++) {
            View listItem = listAdapter.getView(i, null, listView);
            listItem.measure(
                    View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
                    View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
            totalHeight += listItem.getMeasuredHeight();
        }
        ViewGroup.LayoutParams params = listView.getLayoutParams();
        params.height = totalHeight
                + (listView.getDividerHeight() * (listAdapter.getCount() - 1));
        listView.setLayoutParams(params);
    }

    /**
     * 动态设置gridView的高度
     *
     * @param gridView
     * @param columns  列数
     */
    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    public static void setGridViewHeight(GridView gridView, int columns) {
        ListAdapter adapter = gridView.getAdapter();
        int count = adapter.getCount();
        int row = count / columns;
        row = (count % columns) == 0 ? row : (row + 1);
        int totalHeight = 0;
        for (int i = 0; i < row; i++) {
            View view = adapter.getView(i, null, gridView);
            view.measure(View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
                    View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
            totalHeight += view.getMeasuredHeight();
        }
        ViewGroup.LayoutParams layoutParams = gridView.getLayoutParams();
        layoutParams.height = totalHeight + (gridView.getVerticalSpacing() * (row - 1));
        gridView.setLayoutParams(layoutParams);
    }



    /**
     * 获取屏幕宽度
     *
     * @param context
     * @return
     */
    public static int getScreenWidth(Context context) {
        WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display defaultDisplay = wm.getDefaultDisplay();
        return defaultDisplay.getWidth();
    }

    /**
     * 获取屏幕高度
     *
     * @param context
     * @return
     */
    public static int getScreenHeight(Context context) {
        WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display defaultDisplay = wm.getDefaultDisplay();
        return defaultDisplay.getWidth();
    }


    public static int getMatchTitleHeight(Context context) {
        int d = (int) context.getResources().getDisplayMetrics().density;
        int screenHeight = getScreenHeight(context);
        int myheight = 0;
        switch (screenHeight) {
            case 720:
                myheight = 146;
                break;
            case 1080:
                myheight = 300;
                break;
            case 1440:
                myheight = 500;
                break;
            default:
                myheight = 146;

        }
        return myheight;
    }


    /**
     * 从身份证上号码中获取生日
     *
     * @param id
     * @return
     */
    public static long getBirthdayFromID(String id) {
        long birthday = 0;
        if (TextUtils.isEmpty(id) || id.length() < 18) {
            birthday = 0;
        }
        String substring = id.substring(6, 15);
        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd");
        try {
            Date date = sdf.parse(substring);
            birthday = date.getTime();
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return birthday;
    }


    /**
     * 设置状态栏黑色字体图标，
     * 适配4.4以上版本MIUIV、Flyme和6.0以上版本其他Android
     *
     * @param activity
     * @return 1:MIUUI 2:Flyme 3:android6.0
     */
    public static int StatusBarLightMode(Activity activity) {
        int result = 0;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            if (MIUISetStatusBarLightMode(activity.getWindow(), true)) {
                result = 1;
            } else if (FlymeSetStatusBarLightMode(activity.getWindow(), true)) {
                result = 2;
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                activity.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
                result = 3;
            }
        }
        return result;
    }

    /**
     * 已知系统类型时，设置状态栏黑色字体图标。
     * 适配4.4以上版本MIUIV、Flyme和6.0以上版本其他Android
     *
     * @param activity
     * @param type     1:MIUUI 2:Flyme 3:android6.0
     */
    public static void StatusBarLightMode(Activity activity, int type) {
        if (type == 1) {
            MIUISetStatusBarLightMode(activity.getWindow(), true);
        } else if (type == 2) {
            FlymeSetStatusBarLightMode(activity.getWindow(), true);
        } else if (type == 3) {
            activity.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
        }

    }

    /**
     * 清除MIUI或flyme或6.0以上版本状态栏黑色字体
     */
    public static void StatusBarDarkMode(Activity activity, int type) {
        if (type == 1) {
            MIUISetStatusBarLightMode(activity.getWindow(), false);
        } else if (type == 2) {
            FlymeSetStatusBarLightMode(activity.getWindow(), false);
        } else if (type == 3) {
            activity.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_VISIBLE);
        }

    }

    /**
     * 设置状态栏图标为深色和魅族特定的文字风格
     * 可以用来判断是否为Flyme用户
     *
     * @param window 需要设置的窗口
     * @param dark   是否把状态栏字体及图标颜色设置为深色
     * @return boolean 成功执行返回true
     */
    public static boolean FlymeSetStatusBarLightMode(Window window, boolean dark) {
        boolean result = false;
        if (window != null) {
            try {
                WindowManager.LayoutParams lp = window.getAttributes();
                Field darkFlag = WindowManager.LayoutParams.class
                        .getDeclaredField("MEIZU_FLAG_DARK_STATUS_BAR_ICON");
                Field meizuFlags = WindowManager.LayoutParams.class
                        .getDeclaredField("meizuFlags");
                darkFlag.setAccessible(true);
                meizuFlags.setAccessible(true);
                int bit = darkFlag.getInt(null);
                int value = meizuFlags.getInt(lp);
                if (dark) {
                    value |= bit;
                } else {
                    value &= ~bit;
                }
                meizuFlags.setInt(lp, value);
                window.setAttributes(lp);
                result = true;
            } catch (Exception e) {

            }
        }
        return result;
    }

    /**
     * 设置状态栏字体图标为深色，需要MIUIV6以上
     *
     * @param window 需要设置的窗口
     * @param dark   是否把状态栏字体及图标颜色设置为深色
     * @return boolean 成功执行返回true
     */
    public static boolean MIUISetStatusBarLightMode(Window window, boolean dark) {
        boolean result = false;
        if (window != null) {
            Class clazz = window.getClass();
            try {
                int darkModeFlag = 0;
                Class layoutParams = Class.forName("android.view.MiuiWindowManager$LayoutParams");
                Field field = layoutParams.getField("EXTRA_FLAG_STATUS_BAR_DARK_MODE");
                darkModeFlag = field.getInt(layoutParams);
                Method extraFlagField = clazz.getMethod("setExtraFlags", int.class, int.class);
                if (dark) {
                    extraFlagField.invoke(window, darkModeFlag, darkModeFlag);//状态栏透明且黑色字体
                } else {
                    extraFlagField.invoke(window, 0, darkModeFlag);//清除黑色字体
                }
                result = true;
            } catch (Exception e) {

            }
        }
        return result;
    }


    /**
     * 检查设备是否存在SDCard的工具方法
     */
    public static boolean hasSdcard() {
        String state = Environment.getExternalStorageState();
        return state.equals(Environment.MEDIA_MOUNTED);
    }

    /**
     * make true current connect service is wifi
     *
     * @param mContext
     * @return
     */
    public static boolean isWifi(Context mContext) {
        ConnectivityManager connectivityManager = (ConnectivityManager) mContext
                .getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetInfo = connectivityManager.getActiveNetworkInfo();
        if (activeNetInfo != null
                && activeNetInfo.getType() == ConnectivityManager.TYPE_WIFI) {
            return true;
        }
        return false;
    }



    /**
     * 正则表达式校验邮箱
     *
     * @param emaile 待匹配的邮箱
     * @return 匹配成功返回true 否则返回false;
     */
    public static boolean checkEmaile(String emaile) {
        String RULE_EMAIL = "^\\w+((-\\w+)|(\\.\\w+))*\\@[A-Za-z0-9]+((\\.|-)[A-Za-z0-9]+)*\\.[A-Za-z0-9]+$";
        //正则表达式的模式
        Pattern p = Pattern.compile(RULE_EMAIL);
        //正则表达式的匹配器
        Matcher m = p.matcher(emaile);
        //进行正则匹配
        return m.matches();
    }



    /**
     * 把年月日数字变粗体
     */
    public static SpannableString formatYearAndMonthAndDay2Bold(String num) {
        SpannableString spanString = new SpannableString(num);
        StyleSpan span = new StyleSpan(Typeface.BOLD);
        StyleSpan span2 = new StyleSpan(Typeface.BOLD);
        StyleSpan span3 = new StyleSpan(Typeface.BOLD);
        spanString.setSpan(span, 0, 4, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        spanString.setSpan(span2, 5, 7, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        spanString.setSpan(span3, 8, 10, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        return spanString;
    }

}

File: baselib/src/main/java/com/exmple/baselib/utils/CommonUtil.kt
package com.exmple.baselib.utils

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.net.ConnectivityManager
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.support.annotation.RequiresApi
import android.support.v4.view.ViewCompat
import android.text.Spannable
import android.text.SpannableString
import android.text.TextUtils
import android.text.style.StyleSpan
import android.view.*
import android.widget.FrameLayout
import android.widget.GridView
import android.widget.ListView
import com.exmple.baselib.R
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.regex.Pattern

/**
 * Created by mou on 2016/7/6.
 */
object CommonUtil {


    /**
     * 思路是先将statusbar设置成半透明，然后在contentView中添加一个和statusbar一样的高度的view作为statusbar
     * 的背景颜色
     *
     * @param activity
     */
    fun customStatusbar(activity: Activity) {
        val window = activity.window
        val mContentView = activity.findViewById<View>(Window.ID_ANDROID_CONTENT) as ViewGroup

        //First translucent status bar.
        window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        val statusBarHeight = getStatusBarHeight(activity)

        val mChildView = mContentView.getChildAt(0)
        if (mChildView != null) {
            val lp = mChildView.layoutParams as FrameLayout.LayoutParams
            //如果已经为 ChildView 设置过了 marginTop, 再次调用时直接跳过
            if (lp != null && lp.topMargin < statusBarHeight && lp.height != statusBarHeight) {
                //不预留系统空间
                ViewCompat.setFitsSystemWindows(mChildView, false)
                lp.topMargin += statusBarHeight
                mChildView.layoutParams = lp
            }
        }
        // 已经添加过假背景view时，不用再次添加，只需要设置颜色即ok

        var statusBarView: View? = mContentView.getChildAt(0)
        if (statusBarView != null && statusBarView.layoutParams != null && statusBarView.layoutParams.height == statusBarHeight) {
            //避免重复调用时多次添加 View
            statusBarView.setBackgroundColor(activity.resources.getColor(R.color.white))
            //            statusBarView.setBackgroundResource(R.mipmap.ic_launcher);
            return
        }
        statusBarView = View(activity)
        val lp = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, statusBarHeight)
        statusBarView.setBackgroundColor(activity.resources.getColor(R.color.white))
        //        statusBarView.setBackgroundResource(R.mipmap.ic_launcher);
        //向 ContentView 中添加假 View
        mContentView.addView(statusBarView, 0, lp)
    }


    fun fullScreen(activity: Activity) {
        val window = activity.window
        val mContentView = activity.findViewById<View>(Window.ID_ANDROID_CONTENT) as ViewGroup

        //首先使 ChildView 不预留空间
        var mChildView: View? = mContentView.getChildAt(0)
        if (mChildView != null) {
            ViewCompat.setFitsSystemWindows(mChildView, false)
        }

        val statusBarHeight = getStatusBarHeight(activity)
        //需要设置这个 flag 才能设置状态栏
        window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        //避免多次调用该方法时,多次移除了 View
        if (mChildView != null && mChildView.layoutParams != null && mChildView.layoutParams.height == statusBarHeight) {
            //移除假的 View.
            mContentView.removeView(mChildView)
            mChildView = mContentView.getChildAt(0)
        }
        if (mChildView != null) {
            val lp = mChildView.layoutParams as FrameLayout.LayoutParams
            //清除 ChildView 的 marginTop 属性
            if (lp != null && lp.topMargin >= statusBarHeight) {
                //                lp.topMargin -= statusBarHeight;
                mChildView.layoutParams = lp
            }
        }
    }

    fun getStatusBarHeight(activity: Activity): Int {
        val resId = activity.resources.getIdentifier("status_bar_height", "dimen", "android")
        return if (resId > 0) {
            activity.resources.getDimensionPixelSize(resId)
        } else 0
    }

    /**
     * 获取状态栏高度
     *
     * @param context
     * @return
     */
    fun getNavigationBarHeight(context: Context): Int {
        val resources = context.resources
        val resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android")
        return resources.getDimensionPixelSize(resourceId)
    }

    /**
     * 判断手机是否有导航栏
     *
     * @param context
     * @return
     */
    fun hasNavigationBar(context: Context): Boolean {
        //是否在屏幕外有菜单键
        val hasMenuKey = ViewConfiguration.get(context).hasPermanentMenuKey()
        //屏幕外是否有返回键
        val hasBackKey = KeyCharacterMap.deviceHasKey(KeyEvent.KEYCODE_BACK)
        return !hasBackKey && !hasMenuKey

    }

    fun startActivtiy(context: Context, cls: Class<*>) {
        val intent = Intent(context, cls)
        context.startActivity(intent)
    }

    fun startActivtiy(context: Context, cls: Class<*>, bundle: Bundle?) {
        if (bundle == null) {
            startActivtiy(context, cls)
            return
        }
        val intent = Intent(context, cls)
        intent.putExtras(bundle)
        context.startActivity(intent)
    }

    fun startActivtiyForResult(context: Context, cls: Class<*>, requestCode: Int) {
        val intent = Intent(context, cls)
        if (context is Activity) {
            context.startActivityForResult(intent, requestCode)
        } else {
            throw IllegalThreadStateException()
        }
    }

    fun startActivtiyForResult(context: Context, cls: Class<*>, requestCode: Int, bundle: Bundle?) {
        if (bundle == null) {
            startActivtiyForResult(context, cls, requestCode)
            return
        }
        val intent = Intent(context, cls)
        intent.putExtras(bundle)
        if (context is Activity) {
            context.startActivityForResult(intent, requestCode)
        } else {
            throw IllegalThreadStateException()
        }
    }

    /**
     * 判断是否是电话号码
     *
     * @param phoneNum
     * @return
     */
    fun isPhoneNum(phoneNum: String): Boolean {
        return if (TextUtils.isEmpty(phoneNum) || phoneNum.length != 11) {
            false
        } else phoneNum.matches("[1][34578][0-9]{9}".toRegex())
    }


    /**
     * Methods: setListViewHeight Description: 动态获取listview的高度
     *
     * @param listView void
     * @throws null
     */

    fun setListViewHeight(listView: ListView) {
        val listAdapter = listView.adapter ?: return
        var totalHeight = 0
        for (i in 0 until listAdapter.count) {
            val listItem = listAdapter.getView(i, null, listView)
            listItem.measure(
                    View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
                    View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED))
            totalHeight += listItem.measuredHeight
        }
        val params = listView.layoutParams
        params.height = totalHeight + listView.dividerHeight * (listAdapter.count - 1)
        listView.layoutParams = params
    }

    /**
     * 动态设置gridView的高度
     *
     * @param gridView
     * @param columns  列数
     */
    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    fun setGridViewHeight(gridView: GridView, columns: Int) {
        val adapter = gridView.adapter
        val count = adapter.count
        var row = count / columns
        row = if (count % columns == 0) row else row + 1
        var totalHeight = 0
        for (i in 0 until row) {
            val view = adapter.getView(i, null, gridView)
            view.measure(View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
                    View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED))
            totalHeight += view.measuredHeight
        }
        val layoutParams = gridView.layoutParams
        layoutParams.height = totalHeight + gridView.verticalSpacing * (row - 1)
        gridView.layoutParams = layoutParams
    }


    /**
     * 获取屏幕宽度
     *
     * @param context
     * @return
     */
    fun getScreenWidth(context: Context): Int {
        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val defaultDisplay = wm.defaultDisplay
        return defaultDisplay.width
    }

    /**
     * 获取屏幕高度
     *
     * @param context
     * @return
     */
    fun getScreenHeight(context: Context): Int {
        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val defaultDisplay = wm.defaultDisplay
        return defaultDisplay.width
    }


    fun getMatchTitleHeight(context: Context): Int {
        val d = context.resources.displayMetrics.density.toInt()
        val screenHeight = getScreenHeight(context)
        var myheight = 0
        when (screenHeight) {
            720 -> myheight = 146
            1080 -> myheight = 300
            1440 -> myheight = 500
            else -> myheight = 146
        }
        return myheight
    }


    /**
     * 从身份证上号码中获取生日
     *
     * @param id
     * @return
     */
    fun getBirthdayFromID(id: String): Long {
        var birthday: Long = 0
        if (TextUtils.isEmpty(id) || id.length < 18) {
            birthday = 0
        }
        val substring = id.substring(6, 15)
        val sdf = SimpleDateFormat("yyyyMMdd")
        try {
            val date = sdf.parse(substring)
            birthday = date.time
        } catch (e: ParseException) {
            e.printStackTrace()
        }

        return birthday
    }


    /**
     * 设置状态栏黑色字体图标，
     * 适配4.4以上版本MIUIV、Flyme和6.0以上版本其他Android
     *
     * @param activity
     * @return 1:MIUUI 2:Flyme 3:android6.0
     */
    fun StatusBarLightMode(activity: Activity): Int {
        var result = 0
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            if (MIUISetStatusBarLightMode(activity.window, true)) {
                result = 1
            } else if (FlymeSetStatusBarLightMode(activity.window, true)) {
                result = 2
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                activity.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
                result = 3
            }
        }
        return result
    }

    /**
     * 已知系统类型时，设置状态栏黑色字体图标。
     * 适配4.4以上版本MIUIV、Flyme和6.0以上版本其他Android
     *
     * @param activity
     * @param type     1:MIUUI 2:Flyme 3:android6.0
     */
    fun StatusBarLightMode(activity: Activity, type: Int) {
        if (type == 1) {
            MIUISetStatusBarLightMode(activity.window, true)
        } else if (type == 2) {
            FlymeSetStatusBarLightMode(activity.window, true)
        } else if (type == 3) {
            activity.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
        }

    }

    /**
     * 清除MIUI或flyme或6.0以上版本状态栏黑色字体
     */
    fun StatusBarDarkMode(activity: Activity, type: Int) {
        if (type == 1) {
            MIUISetStatusBarLightMode(activity.window, false)
        } else if (type == 2) {
            FlymeSetStatusBarLightMode(activity.window, false)
        } else if (type == 3) {
            activity.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_VISIBLE
        }

    }

    /**
     * 设置状态栏图标为深色和魅族特定的文字风格
     * 可以用来判断是否为Flyme用户
     *
     * @param window 需要设置的窗口
     * @param dark   是否把状态栏字体及图标颜色设置为深色
     * @return boolean 成功执行返回true
     */
    fun FlymeSetStatusBarLightMode(window: Window?, dark: Boolean): Boolean {
        var result = false
        if (window != null) {
            try {
                val lp = window.attributes
                val darkFlag = WindowManager.LayoutParams::class.java
                        .getDeclaredField("MEIZU_FLAG_DARK_STATUS_BAR_ICON")
                val meizuFlags = WindowManager.LayoutParams::class.java
                        .getDeclaredField("meizuFlags")
                darkFlag.isAccessible = true
                meizuFlags.isAccessible = true
                val bit = darkFlag.getInt(null)
                var value = meizuFlags.getInt(lp)
                if (dark) {
                    value = value or bit
                } else {
                    value = value and bit.inv()
                }
                meizuFlags.setInt(lp, value)
                window.attributes = lp
                result = true
            } catch (e: Exception) {

            }

        }
        return result
    }

    /**
     * 设置状态栏字体图标为深色，需要MIUIV6以上
     *
     * @param window 需要设置的窗口
     * @param dark   是否把状态栏字体及图标颜色设置为深色
     * @return boolean 成功执行返回true
     */
    fun MIUISetStatusBarLightMode(window: Window?, dark: Boolean): Boolean {
        var result = false
        if (window != null) {
            val clazz = window.javaClass
            try {
                var darkModeFlag = 0
                val layoutParams = Class.forName("android.view.MiuiWindowManager\$LayoutParams")
                val field = layoutParams.getField("EXTRA_FLAG_STATUS_BAR_DARK_MODE")
                darkModeFlag = field.getInt(layoutParams)
                val extraFlagField = clazz.getMethod("setExtraFlags", Int::class.javaPrimitiveType, Int::class.javaPrimitiveType)
                if (dark) {
                    extraFlagField.invoke(window, darkModeFlag, darkModeFlag)//状态栏透明且黑色字体
                } else {
                    extraFlagField.invoke(window, 0, darkModeFlag)//清除黑色字体
                }
                result = true
            } catch (e: Exception) {

            }

        }
        return result
    }


    /**
     * 检查设备是否存在SDCard的工具方法
     */
    fun hasSdcard(): Boolean {
        val state = Environment.getExternalStorageState()
        return state == Environment.MEDIA_MOUNTED
    }

    /**
     * make true current connect service is wifi
     *
     * @param mContext
     * @return
     */
    fun isWifi(mContext: Context): Boolean {
        val connectivityManager = mContext
                .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetInfo = connectivityManager.activeNetworkInfo
        return if (activeNetInfo != null && activeNetInfo.type == ConnectivityManager.TYPE_WIFI) {
            true
        } else false
    }


    /**
     * 正则表达式校验邮箱
     *
     * @param emaile 待匹配的邮箱
     * @return 匹配成功返回true 否则返回false;
     */
    fun checkEmaile(emaile: String): Boolean {
        val RULE_EMAIL = "^\\w+((-\\w+)|(\\.\\w+))*\\@[A-Za-z0-9]+((\\.|-)[A-Za-z0-9]+)*\\.[A-Za-z0-9]+$"
        //正则表达式的模式
        val p = Pattern.compile(RULE_EMAIL)
        //正则表达式的匹配器
        val m = p.matcher(emaile)
        //进行正则匹配
        return m.matches()
    }


    /**
     * 把年月日数字变粗体
     */
    fun formatYearAndMonthAndDay2Bold(num: String): SpannableString {
        val spanString = SpannableString(num)
        val span = StyleSpan(Typeface.BOLD)
        val span2 = StyleSpan(Typeface.BOLD)
        val span3 = StyleSpan(Typeface.BOLD)
        spanString.setSpan(span, 0, 4, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        spanString.setSpan(span2, 5, 7, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        spanString.setSpan(span3, 8, 10, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        return spanString
    }

}
